File: Stacks Queues/queues/queue.js
class Queue {
    constructor() {
        this.queue = [];
        this.front = -1;
        this.rear = -1;
    }

    isFull() {
        return this.rear === this.queue.length - 1;
    }

    isEmpty() {
        return this.front === -1;
    }

    enqueue(value) {
        if (this.isFull()) {
            console.log("Queue is full");
            return;
        }
        if (this.isEmpty()) {
            this.front = 0;
        }
        this.queue[++this.rear] = value;
        console.log(`Enqueued ${value}`);
    }

    dequeue() {
        if (this.isEmpty()) {
            console.log("Queue is empty");
            return -1;
        }
        const value = this.queue[this.front++];
        if (this.front > this.rear) {
            this.front = this.rear = -1;
        }
        console.log(`Dequeued ${value}`);
        return value;
    }

    peek() {
        if (this.isEmpty()) {
            console.log("Queue is empty");
            return -1;
        }
        return this.queue[this.front];
    }

    printQueue() {
        if (this.isEmpty()) {
            console.log("Queue is empty");
            return;
        }
        console.log("Elements in the queue:");
        for (let i = this.front; i <= this.rear; i++) {
            console.log(this.queue[i]);
        }
    }
}

const queue = new Queue();

// Example usage
queue.enqueue(5);
queue.enqueue(10);
queue.enqueue(15);
queue.printQueue();
queue.dequeue();
queue.printQueue();
console.log("Front element:", queue.peek());

File: Stacks Queues/queues/queue.c
#include <stdio.h>
#define MAX_SIZE 100

int queue[MAX_SIZE];
int front = -1;
int rear = -1;

int isFull() {
    return rear == MAX_SIZE - 1;
}

int isEmpty() {
    return front == -1;
}

void enqueue(int value) {
    if (isFull()) {
        printf("Queue is full\n");
        return;
    }
    if (isEmpty()) {
        front = 0;
    }
    queue[++rear] = value;
    printf("Enqueued %d\n", value);
}

int dequeue() {
    if (isEmpty()) {
        printf("Queue is empty\n");
        return -1;
    }
    int value = queue[front++];
    if (front > rear) {  // Reset front and rear when queue becomes empty
        front = rear = -1;
    }
    printf("Dequeued %d\n", value);
    return value;
}

int peek() {
    if (isEmpty()) {
        printf("Queue is empty\n");
        return -1;
    }
    return queue[front];
}

void printQueue() {
    if (isEmpty()) {
        printf("Queue is empty\n");
        return;
    }
    printf("Elements in the queue:\n");
    for (int i = front; i <= rear; i++) {
        printf("%d\n", queue[i]);
    }
}

int main() {
    int choice, value;

    do {
        printf("Queue Menu:\n");
        printf("1. Enqueue\n");
        printf("2. Dequeue\n");
        printf("3. Peek\n");
        printf("4. Check if queue is empty\n");
        printf("5. Print queue\n");
        printf("6. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                if (isFull()) {
                    printf("Queue is full\n");
                    break;
                }
                printf("Enter value to enqueue: ");
                scanf("%d", &value);
                enqueue(value);
                break;
            case 2:
                dequeue();
                break;
            case 3:
                printf("Front element: %d\n", peek());
                break;
            case 4:
                printf("Is queue empty? %s\n", isEmpty() ? "Yes" : "No");
                break;
            case 5:
                printQueue();
                break;
            case 6:
                printf("Exiting...\n");
                break;
            default:
                printf("Invalid choice\n");
        }
    } while (choice != 6);

    return 0;
}
